Uppy logo upload vue widget

luistoptal opened the pull request:
# Pull request for issue: #376

NOTE: A follow up for this PR implements use of the /tmp folder to store temporary images https://github.com/gigascience/gigadb-website/pull/2092
EDIT: 2092 PR has been merged into this one

This is a pull request for the following functionalities:

* adds uppy uploader to add project logos, with a max height restriction of 60px and max size of 1MB (although it is very unlikely an image of 60px height could reach 1MB size)

## How to test?

Local testing:

### SETUP

- Run anew the up.sh script as it needs to run docker containers for the vue widget
- Modify the autogenerated file `protected/config/yii2/web.php` to use the local file system (this file is reset every time the `up.sh` script is run)

```php
...
        'cloudStore' => [
                   'class' => 'creocoder\flysystem\LocalFilesystem',
                   'path' => '@webroot/files',
                ],
...
```

- in `protected/models/Project.php`, modify the following:

```php
  public static function getStorageBasePath()
  {
      return Yii::getAlias('@web') . '/files'; <-- uncomment this line to use the local file system
      // return 'https://' . self::BUCKET; <-- comment this line
  }
```

NOTE: the above steps need to be done to mock the S3 bucket by the local filesystem. Should be done when running the app locally. I am open to suggestions to improve this flow 

NOTE: unless the user is actively developing the vue widget, the build bundle should be used. In file `protected/views/adminProject/_form.php` I added a $isDev flag that toggles this behavior. If a dev wants to work locally on the vue app, they shoud set it to true. I am open to suggestions to improve this flow. 

### CREATE

- Navigate to http://gigadb.gigasciencejournal.com/adminProject/create
- you should see the uppy dashboard

![image](https://github.com/user-attachments/assets/ea262d3f-6d37-48ea-8806-18656d12df06)

- use the uppy dashboard to upload a file:
  - once you add a file to the dashboard, if the file height is larger than 60px, it should automatically open the file editor, giving you the option to crop the image
- crop the file if needed, and click on the "upload" button
  - if you try to upload a file which is too large, you should see an error message popup and the image should not upload
- the image should be successfully uploaded, and you should see a `files/dev/images/projects/temp/` folder, with a nested uuid folder which contains the image. You should also see the image showing up in the UI

![image](https://github.com/user-attachments/assets/6699a518-35fe-45f5-b11c-ee22b51f6a90)

- fill the new project details and click "create"
- you should be redirected to the view page, looking similar to this
![image](https://github.com/user-attachments/assets/d85d400b-5a1e-4c29-8569-69c8f619151d)
- the `files/dev/images/projects/temp/` folder should be now empty, and there should be a new folder under `files/dev/images/projects/`, looking like a uuid, with the logo within. This uuid is deterministically derived from the newly created project id
- in http://gigadb.gigasciencejournal.com/adminProject/admin you should see the new project listed

### EDIT

- if you try to edit the project:
  - you should see the image previoulsy uploaded
  - you can upload a new image and see how the thumbnail will change

NOTE: as it is expected the logos are quite small, the image is displayed as is, rather than a thumbnail

NOTE: if you try to save the edits, you will see an error informing that duplicate name or url are not valid, which makes it impossible to *just* edit an image (without changing name and url). This is an issue that existed in the preexisting implementation, to reduce the scope of this issue, I suggest to create a new one for this (or fix in a separate PR, for the same issue). For testing purposes changing both fields works

- after saving, the `files/dev/images/projects` should contain, within the same uuid folder, the new file

### DELETE

- Navigate to http://gigadb.gigasciencejournal.com/adminProject/admin
- click on the "delete" icon for the new project
- the deleted project should disappear
- the previous uuid folder should have been deleted

## How have functionalities been implemented?

- a new Vue 3 project was created in the vite folder
- the Vue app uses vite as bundler
- the app is containerized, relevant config in `ops/deployment/docker-compose.yml`, services `vite-project-image-location-prod` and uses a setup similar to the one for the fuw service (`js` service in the docker compose)
- the Vue app is mounted in `protected/views/adminProject/_form.php`, in the element with id `vue-client_project-image-logo`
  - while the vue widget is being mounted, a loading spinner is displayed instead
  - in local dev, the vue app is served in `http://localhost:5173` and the main.ts is loaded directly along with vite client
  - in production, the app is bundled as js scripts and loaded as is (`(cd vite && npm run build)` will create the bundles in `js/vite-logo-upload`)

More instructions about the build flow and testing scripts in `vite/README.md`

Upload flow

1. user adds a file to the uppy dashboard
2. user crops the image as needed to fit dimension requirements
3. user clicks on "upload" button
4. API request to `actionUploadTempLogo` method, that saves the file to a temp folder
5. front receives temp image url and displays the image
6. user fills remaining project data and clicks save
7. API request to `actionCreate` method: A) first, project is created to get a project id, B) then, logo image is stored in a deterministic path (defined by project id), C) finally, project is updated with image url. Additionally, temp url is deleted

If user replaces the logo image for a project, the old image is deleted

NOTE: if user starts the process of uploading a logo to the point where a temp image is created, and they exit the process (reload page, abandon page, etc), the uploaded image will be left dangling in the temp folder. Generally speaking, an image will exist in the temp folder for a brief time, between uploading image with uppy dashboard and clicking on "save" --> this flow is changed in follow up PR https://github.com/gigascience/gigadb-website/pull/2092

## Any issues with implementation?

- I am unable to test with the S3 bucket as I don't have access to it, S3 should be tested by someone who has access to make sure it works, before going to production

## Any changes to automated tests?

Added e2e tests for the logo upload flow in `playwright/tests/admin-project-logo-upload.spec.js`

testing instructions added to `vite/README.md`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Added a Vue-based logo uploader with image editing, validation, and preview for admin project management.
  - Integrated Uppy file uploader for streamlined logo uploads with height and size restrictions.
  - Enhanced admin UI to display uploaded project logos and provide accessible status/error feedback.
  - Project logos are now stored and managed in cloud storage with automatic cleanup on deletion.

- **Bug Fixes**
  - Prevented uploading of images taller than 60px for project logos.

- **Documentation**
  - Added and updated README files with setup, build, and testing instructions for the new Vue client and Playwright tests.
  - Updated changelog to document the new logo uploader feature.

- **Tests**
  - Introduced unit tests for new Vue components and utilities.
  - Added end-to-end tests verifying the admin project logo upload workflow.

- **Chores**
  - Added configuration, ignore, and Docker files to support the new front-end build and deployment process.
  - Updated scripts and deployment routines to include the new logo uploader application.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

luistoptal committed:
feat: setup dockerized vue widget dev server

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
feat: mock image upload

luistoptal committed:
uppy widget is displayed in the ui

luistoptal committed:
display spinner while scripts load

luistoptal committed:
renames controller and model variables to be more specific

luistoptal committed:
feat: a11y uppy image loader

luistoptal committed:
feat: native inplementation

luistoptal committed:
feat: custom image editor

luistoptal committed:
refactor: cleanup current implementation

luistoptal committed:
chore: comment feature and display upload button

luistoptal committed:
feat: able to upload file and send data back to the client

luistoptal committed:
chore: restores actioncreate method to original functionality

luistoptal committed:
feat: use write logo in project controller

luistoptal committed:
feat: setup logo cleanup on project delete and on logo upload

luistoptal committed:
feat: splits vue widget into components

luistoptal committed:
chore: cleanup logs and accept hidden input name

luistoptal committed:
feat: tweak image editor flow and improve a11y

luistoptal committed:
feat: save logo in temp storage

luistoptal committed:
feat: project CRUD wroks with logo

luistoptal committed:
chore: delete logs

luistoptal committed:
feat: setup vue widget dev and prod

luistoptal committed:
refactor: adds global error handling

luistoptal committed:
feat: adds vue component unit testing

luistoptal committed:
feat: validate image height before uploading

luistoptal committed:
chore: use production version of vue app

luistoptal committed:
feat: tweak test scripts

luistoptal committed:
feat: setup prod env

luistoptal committed:
refactor: cleanup vite app dev setup

luistoptal committed:
chor: clean up readme

luistoptal committed:
feat: separates form buttons from uppy widget

luistoptal committed:
chore: reset env

luistoptal committed:
fix: typo

luistoptal committed:
feat: adds e2e test of the logo upload flow

luistoptal committed:
feat: adds tests to readme

luistoptal committed:
chore: expands readme

luistoptal committed:
chore: up log

luistoptal commented:
https://github.com/gigascience/gigadb-website/issues/376 has some interesting comments that should serve to improve this PR

I will apply them

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal commented:
@rija as for this comment: https://github.com/gigascience/gigadb-website/issues/376#issuecomment-2456826542

I think it's better if I implement that in a separate PR, because that is a set of changes I'm not entire sure how it should work, and I want to have it reviewed separately rather than mix with the many changes this PR already has

luistoptal committed:
chore: cleans up

luistoptal committed:
feat: uses tmp folder to store temp uploaded logos in project admin form

luistoptal committed:
chore: remove excessive logs

luistoptal committed:
refactor: simplifies write logo fnc

luistoptal committed:
refactor: update and write tmp logo, set uppy debug to false

rija commented:
Hi @luistoptal,
let me know if this PR depends on other PR, because I'm moving  it up in the PR review column of current sprint board

luistoptal commented:
Hi @rija it doesn't but I would suggest considering this PR and the follow up PR https://github.com/gigascience/gigadb-website/pull/2092 at the same time since they are complementary

the second PR makes use of a tmp folder that allows for easier cleanup of temporary logo images

if a PR depends on any other PR, I always indicate it at the top of the PR body :)

kencho51 reviewed (changes requested):
Hi @luistoptal,

**praise:** I was able to create project image and can see that image is located at `files/dev/images/projects/`
**praise:** The project image can be dragged and dropped into the upload box
**praise:** The project image can be deleted in the adminProject page
**praise:** The size of the image can be adjusted in the upload box
**issue:** When I tried to edit the project image, I got this error:
![image](https://github.com/user-attachments/assets/04171172-d199-4323-b6cf-2141a1a67852)
**suggestion:** For the edit, the name and the url should be allowed to be the same?
**issue:** The vite unit test can only be execute after `npm run dev`, but we want automatic tests
**issue:** The playwright tests `npm run test:single admin-project-logo-upload.spec.js` are failing with these error:
```
Error: Error reading storage state from .auth/admin.json:
ENOENT: no such file or directory, open '.auth/admin.json'
```
**suggestion:** Regarding the above failing tests, is it possible to create a runner you could find in `gigadb-website/tests/`? So the tests can executed automatically, and even can be run in gitlab pipeline
**issue:** There is a need of manual update the in the `/protected/config/yii2/web.php`, which is not preferred
**suggestion:** The update should be done in `ops/configuration/yii-conf/web.*.php.dist`, regarding to different environments.
**issue:** The CloudStore path has to be manually updated to use a local file systems.
**suggestion:** Actually there is already a fake s3 path, which is `fs`, declared in `protected/config/yii2/web.php`, maybe you can have a look at `gigadb-website/fuw/app/backend/models/MoveJob.php` for the usages

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
fix: build playwright auth files before tests

luistoptal commented:
HI @kencho51 thanks for your suggestions and insight

> issue: When I tried to edit the project image, I got this error:
> suggestion: For the edit, the name and the url should be allowed to be the same?

As I mentioned in the PR body, this bug is already present in develop. I proposed to fix it in a separate issue/PR as it is essentially unrelated to the implementation of the logo feature

> issue: The vite unit test can only be execute after npm run dev, but we want automatic tests

I can correctly run the unit tests by 1) Running `./up.sh` script, 2) `cd vite; npm run test:unit` and they pass. `npm run dev` should not be necessary. In principle the only thing needed to run the unit tests would be the node_modules packages and nothing else (`cd vite; npm i; npm run test:unit` should work)

> issue: The playwright tests npm run test:single admin-project-logo-upload.spec.js are failing with these error:

I added a small fix, if you run `npm run test:single logged-in.spec.js` to verify the login I expect it to pass. The `admin-project-logo-upload.spec.js` should pass too.

> suggestion: Regarding the above failing tests, is it possible to create a runner you could find in gigadb-website/tests/? So the tests can executed automatically, and even can be run in gitlab pipeline

Are you referring to the tests folder in the repo?

Not sure how this would work. To set this up, I would possibly expect a playwright step in the ci/cd pipeline, but I am not sure. As far as I see the only thing setup so far is a docker container from which to run the tests manually.

Looking at the existing runners, it seems like the following steps might make sense (I'm not sure):

- add playwright_runner file to the tests folder with the docker compose command:
```bash
set -e
set -u
docker-compose run --rm playwright npm run test
```
- Add it to the `ops/pipelines/gigadb-test-jobs.yml` (?) in the list of scripts

> issue: There is a need of manual update the in the /protected/config/yii2/web.php, which is not preferred
> suggestion: The update should be done in ops/configuration/yii-conf/web.*.php.dist, regarding to different environments.
> issue: The CloudStore path has to be manually updated to use a local file systems.
> suggestion: Actually there is already a fake s3 path, which is fs, declared in protected/config/yii2/web.php, maybe you can have a look at gigadb-website/fuw/app/backend/models/MoveJob.php for the usages

I have been looking into this and suggest some changes, of which I am very hesitant, because I have never dug into this part of the codebase before, so I added them to a new branch, which I will merge into this one if changes look ok. You can review said changes here: https://github.com/luistoptal/gigadb-website/pull/1/files

This kept me running in circles for some time and still don't understand:

- the value of YII_ENV_PROD is truthy in local dev, which seems strange
- Even though GIGADB_ENV == 'dev' is defined in the .env file and it looks like that would determine the value of YII_ENV, YII_ENV == 'prod' in localhost

kencho51 commented:
Hi @luistoptal,


Thanks for the update!

1. I was able to run through all the tests, except:
```
% ./up.sh
% cd vite
% npm run test:unit

> vite-project-uppy-logo-uploader@1.0.0 test:unit
> vitest run


 RUN  v2.1.4 /Users/kencho/gigascience/gigadb-website/vite

 ✓ src/tests/HiddenInput.test.ts (2)
 ✓ src/tests/ImageEditorStatusDisplay.test.ts (7)
 ✓ src/tests/ImageEditorStatusDisplay.type.test.ts (1)
 ✓ src/tests/LogoUploadUppy.test.ts (5)
 ✓ src/tests/UploadedLogoDisplay.test.ts (2)

 Test Files  5 passed (5)
      Tests  17 passed (17)
   Start at  10:14:23
   Duration  550ms (transform 267ms, setup 0ms, collect 695ms, tests 54ms, environment 692ms, prepare 219ms)
% cd ../playwright
% npm run test:single logged-in.spec.js

> e2e@1.0.0 test:single
> npx playwright test --workers=1 --project=chromium logged-in.spec.js


Running 4 tests using 1 worker
  4 passed (8.9s)

To open last HTML report run:

  npx playwright show-report
% npm run test:single tests/admin-project-logo-upload.spec.js 

> e2e@1.0.0 test:single
> npx playwright test --workers=1 --project=chromium tests/admin-project-logo-upload.spec.js


Running 4 tests using 1 worker
  1) [setup] › auth.setup.js:26:6 › authenticate as user ───────────────────────────────────────────

    Error: browser.newContext: Target page, context or browser has been closed




  1 failed
    [setup] › auth.setup.js:26:6 › authenticate as user ────────────────────────────────────────────
  2 skipped
  1 passed (2.4s)

  Serving HTML report at http://localhost:9323. Press Ctrl+C to quit.
^C
```

2. Executing the playwright test in the container seems working, except:
```
% ./tests/playwright_runner
...
 Slow test file: [firefox] › accessibility.spec.js (4.2m)
  Slow test file: [webkit] › accessibility.spec.js (4.0m)
  Slow test file: [chromium] › accessibility.spec.js (3.6m)
  Slow test file: [firefox] › admin-project-logo-upload.spec.js (32.9s)
  Consider splitting slow test files to speed up parallel execution
  10 failed
    [chromium] › accessibility.spec.js:30:9 › Page datasetSubmission/upload › should not have any accessibility violations outside of elements with known issues 
    [chromium] › accessibility.spec.js:30:9 › Page adminRelation/create › should not have any accessibility violations outside of elements with known issues 
    [chromium] › accessibility.spec.js:30:9 › Page adminProject/create › should not have any accessibility violations outside of elements with known issues 
    [firefox] › accessibility.spec.js:30:9 › Page datasetSubmission/upload › should not have any accessibility violations outside of elements with known issues 
    [firefox] › accessibility.spec.js:30:9 › Page adminRelation/create › should not have any accessibility violations outside of elements with known issues 
    [firefox] › accessibility.spec.js:30:9 › Page adminProject/create › should not have any accessibility violations outside of elements with known issues 
    [firefox] › admin-project-logo-upload.spec.js:12:3 › Project logo upload › admin can create and edit project with logo 
    [webkit] › accessibility.spec.js:30:9 › Page datasetSubmission/upload › should not have any accessibility violations outside of elements with known issues 
    [webkit] › accessibility.spec.js:30:9 › Page adminRelation/create › should not have any accessibility violations outside of elements with known issues 
    [webkit] › accessibility.spec.js:30:9 › Page adminProject/create › should not have any accessibility violations outside of elements with known issues 
  334 passed (4.3m)

  Serving HTML report at http://localhost:9323. Press Ctrl+C to quit.
^Cnpm ERR! path /var/www/playwright
npm ERR! command failed
npm ERR! signal SIGINT
npm ERR! command sh -c npx playwright test

npm ERR! A complete log of this run can be found in: /root/.npm/_logs/2025-02-26T03_10_44_053Z-debug-0.log
```

The `playwright_runner` is:
```
set -e
set -u
docker-compose run --rm playwright npm run test
```

3. Regarding adding playwright tests to CICD, you can refer to `acceptance` stage, and create another job, eg. `playwright_gigadb`, for example:

```
playwright_gigadb:
  stage: acceptance 
  paths:
    - playwright/node_modules
    - playwright/tests
  script:
    - ./up.sh
    - cd $APPLICATION
    - ./tests/playwright_runner
   needs: ["sd_gigadb"]
  environment:
    name: dev
```
The above is just a bare minimum template, you may need to further work on it.

Actually, it may not be needed to be run in CICD, as the accessibility seems not a blocking feature for live deployment? Lets wait for @rija and @pli888 comments first.

4. Regarding your new branch https://github.com/luistoptal/gigadb-website/pull/1/files, I think:
  i. You dont need to create new `ops/configuration/yii-conf/web.dev.php.dist` and refactor the `generata_config.sh`
  ii. And refactor the function `getStorageBasePath()` likes this:
```
public static function getStorageBasePath()
  {
      if (in_array($_SERVER['GIGADB_ENV'], ["dev","CI"])) {
          return Yii::$app->fs;
      } else {
          return Yii::$app->CloudStore;
      }
  }
```

The above is just pseudo code, further work on implementation will be needed.

luistoptal committed:
chore: run anpm audit fix on vite project

luistoptal commented:
Hi @kencho51 thanks for your response,

> i. You dont need to create new ops/configuration/yii-conf/web.dev.php.dist and refactor the generata_config.sh
> ii. And refactor the function getStorageBasePath() likes this

This approach would involve checking environment in both protected/models/Project.php and  protected/controllers/AdminProjectController.php because both access the cloud storage

This also mixes business logic and config logic. Wouldn't it make more sense to define the store in the config rather than the controller and model files?

@rija @kencho51 do you know if the CI/CD environment is supposed to be able to access the S3 bucket? Given what I see, I would guess that the answer is yes but could you confirm?

kencho51 commented:
Hi @luistoptal,

> @rija @kencho51 do you know if the CI/CD environment is supposed to be able to access the S3 bucket? Given what I see, I would guess that the answer is yes but could you confirm?

Our CICD is able to access the s3 bucket, details you can look at `gigadb-website/tests/acceptance/Wasabi.feature' and
`gigadb-website/tests/_support/Steps/DeveloperSteps.php`

> This approach would involve checking environment in both protected/models/Project.php and protected/controllers/AdminProjectController.php because both access the cloud storage 
> 
> This also mixes business logic and config logic. Wouldn't it make more sense to define the store in the config rather than the controller and model files?

Im not sure how exactly it should be implemented, maybe wait for @rija and @pli888 comments first.

rija reviewed (approved):
Hi @luistoptal,

I've started looking into this PR.
I haven't looked at the code in details, but I looked at the functionality.

just to get it out the way, the duplicate fields issue when updating a project is distracting, but there is a ticket (#1290) and a PR (#1924) already for that.
So I'm now reviewing that PR (#1924) so that I can continue later review this PR without that distraction. What's below is my first round of observations. I'll come back to this PR over a few rounds.

(you can follow state of all PRs here: https://github.com/orgs/gigascience/projects/90/views/4)

functionality wise (local deployment):

**praise**: I can create a new project with a new image
**praise**: I could resize the image using the embedded editor

**issue**: if the image I choose (most of the examples I had were like that unfortunately) has too big dimensions (even though is under the file size limit), the embedded editor will refuse to reduce the height to 60px of less, the resize tool will lock before that. I had to downsize my examples images using other software first (several times as I didn't know what the threshold of acceptance was).

**issue**: When I view the project record in the admin dashboard, I expected to see the uploaded image there, but It doesn't seem to be the case

**issue**: I reckon there are too many steps involved to upload the image: actual file selection by the user, eventual resizing, clicking save, clicking upload, clicking done, and then clicking create
**suggestion**: I wonder is of some of those steps could happen automatically.

**issue**: When deleting a project record from admin dashboard I get an error _"Invalid request. Please do not repeat this request again."_
**suggestion**: IIRC it has to do with the REST HTTP verb used to send the delete request. It's trying to do a DELETE using a GET query or something like that.

**nitpick**: I think the curators want the project height to be exactly 60px ideally, so it would be useful to have a hint whenever our resizing bring the height to 60px in the embedded editor, like a green version of the red hint that will only appear when we hit 60px from above or from below

**praise**: I like the visual hint for when the image height is still too big

Implementation wise (didn't look into the details yet, this is based on skimming the code and reading the exchange between you and @kencho51):

**issue**: As it stands, even if there were no errors, we cannot merge this PR because we can't reference the images from the server directory.
I know you created a separate PR #2092 to make it easier to reason the work, but I really think you should integrate in this PR the work done in #2092.

**note**: For local testing/development vs production deployment, you don't need to alter the code. There are already two Flysystem application components configured by default to that effect. `localStore` is for local filesystem (That's the one you should have used in your first implementation, it would have avoided manual code change as setup), while `cloudStore` is for S3.

However, using the latter exclusively doesn't make it easy to have a temporary storage on server that the JS editor can access, 
but you can combine both using the Flysystem mount manager.
Use `localStore` to handle temporary files on the server.
Then `cloudStore` to upload edited image to S3 (*)

**suggestion**: Docs here: https://flysystem.thephpleague.com/v1/docs/advanced/mount-manager/

**note**:(*) You should be able to access S3 from anywhere, even from local dev, as long as you have the right AWS API keys.
The important part is to make sure you use the right environment-based directory: 
The S3 bucket of interest (`assets.gigadb-cdn.net`) has the following top-level directories:
`CI`, `dev`, `live`, `staging`.
Make sure that the code writes the images inside the correct directory that matches the environment it is deployed to (including local dev).

rija reviewed (changes requested):
Hi @luistoptal,

I've started looking into this PR.
I haven't looked at the code in details, but I looked at the functionality.

just to get it out the way, the duplicate fields issue when updating a project is distracting, but there is a ticket (#1290) and a PR (#1924) already for that.
So I'm now reviewing that PR (#1924) so that I can continue later review this PR without that distraction. What's below is my first round of observations. I'll come back to this PR over a few rounds.

(you can follow state of all PRs here: https://github.com/orgs/gigascience/projects/90/views/4)

functionality wise (local deployment):

**praise**: I can create a new project with a new image
**praise**: I could resize the image using the embedded editor

**issue**: if the image I choose (most of the examples I had were like that unfortunately) has too big dimensions (even though is under the file size limit), the embedded editor will refuse to reduce the height to 60px of less, the resize tool will lock before that. I had to downsize my examples images using other software first (several times as I didn't know what the threshold of acceptance was).

**issue**: When I view the project record in the admin dashboard, I expected to see the uploaded image there, but It doesn't seem to be the case

**issue**: I reckon there are too many steps involved to upload the image: actual file selection by the user, eventual resizing, clicking save, clicking upload, clicking done, and then clicking create
**suggestion**: I wonder is of some of those steps could happen automatically.

**issue**: When deleting a project record from admin dashboard I get an error _"Invalid request. Please do not repeat this request again."_
**suggestion**: IIRC it has to do with the REST HTTP verb used to send the delete request. It's trying to do a DELETE using a GET query or something like that.

**nitpick**: I think the curators want the project height to be exactly 60px ideally, so it would be useful to have a hint whenever our resizing bring the height to 60px in the embedded editor, like a green version of the red hint that will only appear when we hit 60px from above or from below

**praise**: I like the visual hint for when the image height is still too big

Implementation wise (didn't look into the details yet, this is based on skimming the code and reading the exchange between you and @kencho51):

**issue**: As it stands, even if there were no errors, we cannot merge this PR because we can't reference the images from the server directory.
**suggestion**:I know you created a separate PR #2092 to make it easier to reason the work, but I really think you should integrate in this PR the work done in #2092.

**note**: For local testing/development vs production deployment, you don't need to alter the code. There are already two Flysystem application components configured by default to that effect. `localStore` is for local filesystem (that's the one you could used in your first implementation, it would have avoided manual code change as setup), while `cloudStore` is for S3.

However, using the latter exclusively doesn't make it easy to have a temporary storage on server that the JS editor can access, 
but you can combine both using the Flysystem mount manager.
Use `localStore` to handle temporary files on the server.
Then `cloudStore` to upload edited image to S3 (*)

**suggestion**: Docs here: https://flysystem.thephpleague.com/v1/docs/advanced/mount-manager/

**note**:(*) You should be able to access S3 from anywhere, even from local dev, as long as you have the right AWS API keys.
The important part is to make sure you use the right environment-based directory: 
The S3 bucket of interest (`assets.gigadb-cdn.net`) has the following top-level directories:
`CI`, `dev`, `live`, `staging`.
Make sure that the code writes the images inside the correct directory that matches the environment it is deployed to (including local dev).

rija commented:
Hi @luistoptal, 

For info, #1924 has now been merged.
You probably want to rebase and fix conflicts

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
chore: Merge branch 'new-feature/376-upload-project-logos-tmp-folder' into new-feature/376-upload-update-project-images

luistoptal committed:
chore: Configure CORS for Vite development server

luistoptal committed:
chore: remove redundant changelog entry

luistoptal committed:
feat: Enhance project view with image display and formatting

luistoptal committed:
feat: display logo image in project admin page

luistoptal committed:
feat: streamline logo upload steps

luistoptal committed:
feat: granular crop height message, migrate less variables

luistoptal commented:
Hi @rija I still need to wrap my head around the part of the Flysystem stores

As for the rest:

> should integrate in this PR the work done in #2092

You are right, the first I have done is merge #2092


> embedded editor will refuse to reduce the height to 60px of less

you can click on the zoom in button and you can also zoom with the mouse whell, that will make it possible to reach 60px height even for large images. As for resizing the image instead of cropping, I tried when I originally worked on this feature but that is not possible with uppy / cropper as far as I know

Maybe I can add a description under the widget to offer contextual help mentioning the zoom option

> When deleting a project record from admin dashboard I get an error "Invalid request. Please do not repeat this request again."


I assume you mean clicking delete in one project in this screen: http://gigadb.gigasciencejournal.com/adminProject/admin

I cannot reproduce this in dev env. The network tab shows POST request on delete. I think it might be related to the `beforeDelete` method in the Project model. I will take a look at this again once everything else has been resolved

> When I view the project record in the admin dashboard, I expected to see the uploaded image there, but It doesn't seem to be the case

I updated the view and admin views to render the images instead. I also made the image column not sortable since I think sorting by image is strange

![image](https://github.com/user-attachments/assets/ec946c3d-8fc4-48f9-853a-88bdf5e45763)
![image](https://github.com/user-attachments/assets/c7ef97d7-d610-4272-8cc3-b753649a6aed)


> too many steps

I made some changes and simplified the process as much as possible minimizing the number of steps to add a logo to the bare minimum

> green version of the red hint that will only appear when we hit 60px from above or from below

Added more nuanced message on crop resize

rija reviewed (changes requested):
Hi @luistoptal,

thank you for your reply.

**thought**: Let me think about the image pipeline more and see if we can separate that part from the UI part into a separate ticket and maybe have a backend engineer look at that one, especially given there is more backend work to deal with image deletion (see my comment further down)

**praise**: upload of project image seems streamlined now

**issue**: cropping limit: I didn't used the "zoom in" function before, but now I have tried again and try to use it, and it still won't allow me to crop all the way to 60px with big-ish images and I have tried with three different browsers. 
Here's the video of my struggling in Firefox: 
https://drive.google.com/file/d/1-GXyo3iMkVcAUUR7NLSNmliUFfiV0AU3/view?usp=sharing

**praise**: I can delete project now, maybe it was a just a spurious error
**question**: what happened to the associated image when a project is deleted?
**note**: we have a database table called `images_todelete`, we use it when a curators delete the main image for a dataset. I think it's also suitable for deleting project image, so that when a project is deleted, the image location is added to that table and a houskeeping job will later asynchronously remove files listed in thattable, but it might not be the right ticket for that. I may create a separate ticket, and it maybe a backend work only anyways.

**issue**: I cannot update most existing projects, because I've got a new error: 
```
Please fix the following input errors:

- Image Location is too long (maximum is 100 characters).
```

It sounds like a bug hidden behind the one we've just fixed. 

**suggestion**: The fix is easy though:
In `protected/models/Project.php`,
replace on line 67:

`array('image_location', 'length', 'max'=>100),`

with: 

`array('image_location', 'length', 'max'=>255),`

to match the size of the database column.
(fortunately it seems like the autogenerated random string based location is still within that range?)


**issue**: didn't see any difference in nuance for the height hint, on Firefox and Safari and after hard-resetting cache.

**praise**: Nice to be able to view the image in the project view and project list

luistoptal committed:
fix: Increase project image location max length to 255 characters

luistoptal commented:
Hi @rija 

> won't allow me to crop all the way to 60px

For the case you show in the video there are two options:

1) we would need to somehow implement a resizing feature (more complicated than cropping)
2) the user would need to screenshot or resize the logo to an appropriate size outside the application

It would be possible to keep zooming in until achieving the right height but unavoidably in the case in the video that would mean to cut off the unicorn logo

> what happened to the associated image when a project is deleted?

It should be deleted too. More specifically:

- the temporary logo (after upload step, but before creating / updating project) should be added to a /tmp folder in the server local store, which I believe should be cleaned up automatically from time to time routinely by default
- once the project is created, the logo should be persisted in the S3 bucket. If the project is deleted, then the `beforeDelete` hook should cleanup the logo
- if the logo is replaced by a new one, in the actionUpdate method a call to deleteLogo should delete the old one

I assume this is correct, but please let me know if it should be different

> Image Location is too long

Applied your suggested fix

> didn't see any difference in nuance for the height hint

this is what is expected now in the UI:

![image](https://github.com/user-attachments/assets/1c094e62-bcd6-4994-bbad-91e9e3d0da58)
![image](https://github.com/user-attachments/assets/30b091a1-9341-4e2f-8c25-e42c81d98e16)
![image](https://github.com/user-attachments/assets/002a626d-3da4-4576-a27a-335d1e52957f)

it is only displayed during cropping when changing the height between too tall / too short

It can be tested with e.g. https://placehold.co/200x90.png

> Let me think about the image pipeline more

I was thinking about how to manage the storages, and getting a bit stuck. I will stop for now, and wait for your decision

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
chore: skipping outdated test

luistoptal commented:
Updated the code to skip a functional test which was outdated, expecting to find the removed image location input

I simply marked it as skipped, could also be removed

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal commented:
@coderabbitai full review

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
fix: update Playwright test commands and improve image loading error handling

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal commented:
merged develop

rija reviewed (changes requested):
Hi @luistoptal,

**praise**: Interactive cropping of image in uploader works
**praise**: testing locally, creating, changing image, updating details and deleting project works

issue: the instructions for getting it work locally is a hack and doesn't leverage the fact that the local filesystem for Flysystem is already configured
suggestion: A better way to handle the two modes is to use both store types and detect in code which environment we are in.

eg: 

```
if ("dev" == Yii::$app->params['environment']) {
//do the stuff specific to dev environment
 }
else {
// cloud version
}
```

if you look at a `protected/config/yii2/web.php`

there is already a `fs` type of storage already configured

so the code in `protected/models/Project.php`

will looks like something like this in `protected/models/Project.php`:

```
public static function getStorageBasePath()  
    {  
        if ("dev" == Yii::$app->params['environment']) {  
            return Yii::getAlias('@web') . '/files';  
        }  
        else {  
            return 'https://' . self::BUCKET;  
        }
```


and  in `protected/controllers/AdminProjectController.php`

```
public function actionCreate()  
    {  
         $model = new Project;  
      
         // Uncomment the following line if AJAX validation is needed  
         // $this->performAjaxValidation($model);  
      
         if(isset($_POST['Project']))  
         {  
             $model->attributes = $_POST['Project'];  
             $storage = Yii::$app->cloudStore;
						   if ("dev" == Yii::$app->params['environment']) {  
                $storage = Yii::$app->fs;  
              }
					...
        } 

```



**issue**: on my AWS deployment, the Uppy app doesn't appear and this error shows up in JS console:
```
Refused to load data:,(function t(){"undefined"==typeof globalThis&&(window.globalThis=window);const e=globalThis.document,t=globalThis.CustomEvent,n=globalThis.FileReader,r=globalThis.Blob,a={family:"font-family",style:"font-style",weight:"font-weight",stretch:"font-stretch",unicodeRange:"unicode-range",variant:"font-variant",featureSettings:"font-feature-settings"};if(globalThis.FontFace){const n=globalThis.FontFace;globalThis.FontFace=function(){return o(...arguments).then((n=>e.dispatchEvent(new t("single-file-new-fon...de] }"};const a=e.fonts.clear;e.fonts.clear=function(){return e.dispatchEvent(new t("single-file-clear-fonts")),a.call(e.fonts)},e.fonts.clear.toString=function(){return"function clear() { [native code] }"}}async function o(e,t,o){const s={};return s["font-family"]=e,s.src=t,o&&Object.keys(o).forEach((e=>{a[e]&&(s[a[e]]=o[e])})),new Promise((e=>{if(s.src instanceof ArrayBuffer){const t=new n;t.readAsDataURL(new r([s.src])),t.addEventListener("load",(()=>{s.src="url("+t.result+")",e(s)}))}else e(s)}))}})() because it does not appear in the script-src directive of the Content Security Policy.
```

**suggestion**: if this code is from the built vue.js app, may be the path to it  should be added to the CSP directive in the Nginx config

**issue**: In the pipeline two acceptance test scenario are failing. One seems unrelated, the other is related: 
```
There were 2 failures:

1) form to update dataset details: Open private url is working

Test tests/acceptance/AdminDatasetForm.feature:Open private url is working

Step  I should see current url contains "/dataset/100039/token/" 

Fail  Failed asserting that '/adminDataset/update/id/5' contains "/dataset/100039/token/".

Scenario Steps:

18. $I->seeInCurrentUrl("/dataset/100039/token/") at tests/_support/AcceptanceTester.php:356

17. $I->wait("1") at tests/_support/AcceptanceTester.php:236

16. $I->click("Create/Reset Private URL") at tests/_support/AcceptanceTester.php:194

15. $I->amOnPage("/adminDataset/update/id/5") at tests/_support/AcceptanceTester.php:39

14. $I->waitForText("Home",10) at tests/_support/Steps/CuratorSteps.php:140

13. $I->click("Login") at tests/_support/Steps/CuratorSteps.php:139

Artifacts:

Png: /project/tests/_output/AdminDatasetForm.Open.private.url.is.working.fail.png

Html: /project/tests/_output/AdminDatasetForm.Open.private.url.is.working.fail.html

2) form to manage project: Can update image location

Test tests/acceptance/AdminProject.feature:Can update image location

Step  I fill in the field of "name" "project[image_location]" with "<https://test/images/projects/genome_10k/g10>](https://test/images/projects/genome_10k/g10) klogo.jpg" 

Fail  element with name 'Project[image_location]' was not found.

Scenario Steps:

14. $I->fillField({"name":"Projec...},"<https://test/i>](https://test/i)...") at tests/_support/AcceptanceTester.php:170

13. $I->amOnPage("/adminProject/update/id/2") at tests/_support/AcceptanceTester.php:39

12. $I->waitForText("Home",10) at tests/_support/Steps/CuratorSteps.php:140

11. $I->click("Login") at tests/_support/Steps/CuratorSteps.php:139

10. $I->fillField({"name":"LoginForm[password]"},"gigadb") at tests/_support/Steps/CuratorSteps.php:138

9. $I->fillField({"name":"LoginForm...},"admin@gigadb.org") at tests/_support/Steps/CuratorSteps.php:137
```
suggestion: check whether they pass for you locally and on pipeline

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

luistoptal committed:
feat: adds switch for storages

luistoptal commented:
HI @rija , thanks for your review

manually running this locally I am finding one issue with the process of updating the logo in an existing project, I will be looking into it

I am also trying to deploy my fork to staging

As for the failing test; it should be either removed or replaced because it is testing the old functionality. Playwright tests test the full logo workflow, but I imagine it would be more straightforward to rely on acceptance tests following specs of the rest of the site. I will be reviewing that too

luistoptal committed:
refactor: common get storage function for controller and model

luistoptal committed:
fix: solves e2e test failures

luistoptal committed:
fix: update acceptance test so that it passes with new logo widget

luistoptal commented:
@rija I have updated / fixed all the issues except the ones related to staging, I will be checking this once / if my pipeline succeeds

luistoptal committed:
chore: Merge branch 'develop' into new-feature/376-upload-update-project-images

rija reviewed (changes requested):
Hi @luistoptal 

**praise**: all tests are passing
**praise**: still work on local dev with no hacks
**issue**: Vue app not appearing on AWS deployment like before
It's not clear it's a CSP issue anymore as the previously mention CSP error doesn't appear anymore in the console
On some browsers there is a spinning wheel that keep spinning instead of of showing the file upload app
**suggestion**: one of the dev will look further into this issue

kencho51 commented:
Hi,

> **suggestion**: one of the dev will look further into this issue
I confirm that the Vue app is not appearing on my staging neither when I go to `/adminProject/create` or `/adminProject/update/id/51`, the only thing I could see is the spinning wheel.
<img width="1414" height="601" alt="image" src="https://github.com/user-attachments/assets/89482c33-02c5-4b91-9286-4aae4c078424" />

**issue:** Actually, there is a CSP related problem in the `/adminProject/admin`, some of the image in the image column cannot be deisplayed:
<img width="1422" height="1017" alt="image" src="https://github.com/user-attachments/assets/97fbceb4-e104-4b7d-a677-d17a845f01e8" />
And here is the console log:
```
admin:352 Refused to load the image 'https://github.com/FAANG/comm-data-portal-projects/raw/master/projects/faang/project-logo.png' because it violates the following Content Security Policy directive: "img-src blob: data: 'self' https://assets.gigadb-cdn.net *.tile.openstreetmap.org".
admin:352 Refused to load the image 'https://www.mpi-cbg.de/fileadmin/_processed_/5/d/csm_header-EUROFish2_b9875d945f.jpg' because it violates the following Content Security Policy directive: "img-src blob: data: 'self' https://assets.gigadb-cdn.net *.tile.openstreetmap.org".
``` 

And by looking into the code, the whole `vite` project is setup in the docker-compose.yml, which I think the production docker-compose file need to updated and the the `vite` dir needs to be baked into the docker file for the production.

rija commented:
Hi @kencho51, @luistoptal,

Regarding the CSP issue for the images, this is now a legacy data issue:
This PR will ensure the images for project are stored in our system, instead of being deep links to other websites.
That means that this PR should be accompanied by a migration step where all existing project images are be re-uploaded.
